### PR TITLE
[Codec] Finite scalar quantizer

### PR DESCRIPTION
**Pull request by anteju**
# What does this PR do ?

This PR adds finite scalar quantizer as used in
Mentzer et al., Finite Scalar Quantization: VQ-VAE Made Simple (https://arxiv.org/abs/2309.15505v1)

**Collection**: TTS

# Changelog 
- Added `FiniteScalarQuantizer`
- Added unit tests

# Usage
This can be used as a drop-in replacement for `ResidualVectorQuantizer`, for example using the following configuration

```yaml
model:
  ...
  commit_loss_scale: 0.0 # need to be set to zero, since FSQ does not include commit loss
  ...
  vector_quantizer:
    _target_: nemo.collections.tts.modules.audio_codec_modules.FiniteScalarQuantizer
    num_levels: [8, 5, 5, 5]
```

For using larger vector, `GroupFiniteScalarQuantizer` can be used to repeat quantization pattern several times, for example using the following configuration

```yaml
model:
  ...
  commit_loss_scale: 0.0 # need to be set to zero, since FSQ does not include commit loss
  ...
  vector_quantizer:
    _target_: nemo.collections.tts.modules.audio_codec_modules.GroupFiniteScalarQuantizer
    num_groups: 3
    num_levels_per_group: [3, 4]
```

This corresponds to FSQ with `num_levels: [3, 4, 3, 4, 3, 4]`.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: tests/collections/tts/modules/test_audio_codec_modules.py [hunk 1]
```diff
@@ -205,3 +207,183 @@ def test_snake(self):
...
+        """
+        num_levels = [2, 3]
+        fsq = FiniteScalarQuantizer(num_levels=num_levels)
+
+        # # To generate inputs & outputs for testing
+        # max_len = 8
+        # inputs = torch.randn([self.num_examples, fsq.codebook_dim, max_len])
+        # input_len = torch.tensor([max_len] * self.num_examples, dtype=torch.int32)
+        # dequantized, indices = fsq(inputs=inputs, input_len=input_len)
+        # print(inputs)
```
Comment (rlangman) [inline tests/collections/tts/modules/test_audio_codec_modules.py:259]: Remove commented code

File: tests/collections/tts/modules/test_audio_codec_modules.py [hunk 2]
```diff
@@ -205,3 +207,183 @@ def test_snake(self):
...
+        inputs = torch.tensor(
+            [
+                [
+                    [0.6572, 1.3574, 2.1646, 0.9457, -0.3489, 0.6732, -0.7148, -0.5143],
+                    [0.5123, 0.8552, 1.7814, 1.9938, -1.1909, -0.9991, -3.7932, -0.4438],
+                ],
+                [
+                    [0.2357, 0.8324, 0.8932, -0.0596, 0.6130, -0.0299, 0.3824, 1.6278],
+                    [-0.3781, 0.1864, -0.2190, 1.2199, -1.1398, -0.8443, -0.7865, 0.1470],
+                ],
```
Comment (rlangman) [inline tests/collections/tts/modules/test_audio_codec_modules.py:274]: Should we use smaller dimension inputs/outputs for the tests? 80 elements looks unnecessarily big for a reader to try and parse and validate for correctness.

File: nemo/collections/tts/models/audio_codec.py [hunk 3]
```diff
@@ -77,6 +77,16 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
             self.vector_quantizer = instantiate(cfg.vector_quantizer)
+
+            vq_output_types = list(self.vector_quantizer.output_types.keys())
+
+            if len(vq_output_types) == 3 and vq_output_types[-1] == 'commit_loss':
+                self.vector_quantizer_has_commit_loss = True
+                logging.info('Vector quantizer supports commit loss.')
+            else:
+                self.vector_quantizer_has_commit_loss = False
+                logging.info('Vector quantizer does not support commit loss.')
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/models/audio_codec.py:88]: If it is not much extra work, I think having the abstract base class would be better. It does not matter much now with only 2 implementations, but should help long term as we inevitably add more types of quantizers.

**Review by rlangman — APPROVED** [review]
LGTM

**Review by nithinraok — APPROVED** [review]
LGTM

File: nemo/collections/tts/models/audio_codec.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/models/audio_codec.py:88]: Sure, sounds good.

**Pull request merged**